lib/pka_dev.c

```diff
@@ -1928,24 +1940,43 @@ int __pka_dev_open_ring(uint32_t ring_id)
         return -EPERM;
 
     ring = pka_dev_get_ring(ring_id);
+
+    mutex_lock(&ring->mutex);
+
     if (!ring || !ring->shim)
+    {
+        mutex_unlock(&ring->mutex);
         return -ENXIO;
+    }
+
+    if (ring->status == PKA_DEV_RING_STATUS_BUSY)
+    {
+        mutex_unlock(&ring->mutex);
+        return -EBUSY;
+    }
 
     shim = ring->shim;
 
     if (shim->status == PKA_SHIM_STATUS_UNDEFINED ||
           shim->status == PKA_SHIM_STATUS_CREATED ||
           shim->status == PKA_SHIM_STATUS_FINALIZED)
+    {
+        mutex_unlock(&ring->mutex);
         return -EPERM;
+    }
 
     if (ring->status != PKA_DEV_RING_STATUS_INITIALIZED)
+    {
+        mutex_unlock(&ring->mutex);

```

Comment:
I suggest you create a tag and jump into it if the statements are not verified, for example:

  if (condition1)
  {
      ret = -EPERM;
      goto done;
  }

  ...

  done:
     mutex_unlock(&ring->mutex);
     return ret;